File: 042/cpp/main.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int i = 0;
        int j = height.size() - 1;
        int maxLeft = 0;
        int maxRight = 0;
        int n = 0;
        while (i < j) {
            if (height[i] <= height[j]) {
                maxLeft = max(maxLeft, height[i]);
                n += maxLeft - height[i];
                i++;
            } else {
                maxRight = max(maxRight, height[j]);
                n += maxRight - height[j];
                j--;
            }
        }
        return n;
    }
};

File: 042/rust/main.rs
use std::cmp;

impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let mut i = 0;
        let mut j = height.len() - 1;
        let mut maxLeft = 0;
        let mut maxRight = 0;
        let mut n = 0;
        while i < j {
            if height[i] < height[j] {
                maxLeft = cmp::max(maxLeft, height[i]);
                n += maxLeft - height[i];
                i += 1;
            } else {
                maxRight = cmp::max(maxRight, height[j]);
                n += maxRight - height[j];
                j -= 1;
            }
        }
        return n;
    }
}
